lib/ReadsUtils/ReadsUtilsImpl.py

```diff
@@ -106,39 +111,92 @@ def _proc_upload_reads_params(self, params):
         if not self.xor(objid, name):
             raise ValueError(
                 'Exactly one of the object ID or name must be provided')
-        revid = params.get('rev_id')
-        revfile = params.get('rev_file')
-        if revid and revfile:
-            raise ValueError('Specified both a local file and a shock node ' +
-                             'for the reverse reads file')
-        if shock and revfile:
-            raise ValueError('Cannot specify a local reverse reads file ' +
-                             'with a forward reads file in shock')
-        if not shock and revid:
-            raise ValueError('Cannot specify a reverse reads file in shock ' +
-                             'with a local forward reads file')
-        if not shock and revfile:
-            revid = os.path.abspath(os.path.expanduser(revfile))
+
+        revsource = self._process_rev_params(params.get('rev_id'), params.get('rev_file'), 
+            params.get('rev_file_url'), params.get('rev_staging_file_name'), reads_source)
+
         interleaved = 0
         kbtype = 'KBaseFile.SingleEndLibrary'
         single_end = True
-        if params.get('interleaved') or revid:
+        if params.get('interleaved') or revsource:
             interleaved = 1
             kbtype = 'KBaseFile.PairedEndLibrary'
             single_end = False
 
         source_reads_ref = params.get('source_reads_ref')
         if source_reads_ref:
-            o = self._propagate_reference_reads_info(params, dfu, source_reads_ref,
+            o = self._propagate_reference_reads_info(params, dfu,
+                                                     source_reads_ref,
                                                      interleaved, single_end)
         else:
             o = self._build_up_reads_data(params, single_end)
-        return o, wsid, name, objid, kbtype, single_end, fwdid, revid, shock
+        return o, wsid, name, objid, kbtype, single_end, fwdsource, revsource, reads_source
+
+    def _check_rev_params(self, revid, revfile, revurl, revstaging, reads_source):
+        if sum(bool(e) for e in [revid, revfile, revurl, revstaging]) > 1:
+            raise ValueError('Cannot specify more than one rev file source')
+
+        if reads_source == 'shock' and any([revfile, revurl, revstaging]):
```

Comment:
👍  removed this block.

Next block actually is tested in test_upload_fail_spec_fwd_file_rev_id.